File: neuro-backend/src/messaging/consumer.ts
import amqp, { type ConsumeMessage } from 'amqplib'
import { app } from '../app.js'
import { env } from '../env/index.js'

const QUEUE_ARGS = {
    'x-dead-letter-exchange': 'neurofinance.dlx',
}

function headerValue(headers: Record<string, unknown> | undefined, name: string): string | undefined {
    if (!headers) {
        return undefined
    }
    const value = headers[name] ?? headers[name.toLowerCase()]
    if (Array.isArray(value)) {
        return value[0] ? String(value[0]) : undefined
    }
    return value ? String(value) : undefined
}

async function handleMessage(raw: ConsumeMessage) {
    const request = JSON.parse(raw.content.toString()) as {
        method?: string
        path?: string
        payload?: unknown
        headers?: Record<string, unknown>
    }

    const method = (request.method || 'GET').toUpperCase()
    const path = request.path || '/'
    const authorization = headerValue(request.headers, 'authorization')
    const injectHeaders: Record<string, string> = {}
    if (authorization) {
        injectHeaders.authorization = authorization
    }

    const hasBody = request.payload !== undefined && request.payload !== null && method !== 'GET' && method !== 'DELETE'
    if (hasBody) {
        injectHeaders['content-type'] = 'application/json'
    }

    const response = await app.inject({
        method: method as 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE',
        url: path,
        headers: injectHeaders,
        ...(hasBody ? { payload: request.payload as object } : {}),
    })

    const contentType = String(response.headers['content-type'] ?? '')
    let body: unknown = response.body
    if (contentType.includes('application/json') && response.body) {
        try {
            body = JSON.parse(response.body)
        } catch {
            body = response.body
        }
    }

    return { status: response.statusCode, body }
}

export async function startBackendConsumer() {
    if (env.MESSAGING !== 'amqp') {
        console.log('[rabbitmq] MESSAGING=http — consumidor do backend desligado')
        return
    }
    if (!env.RABBITMQ_URL) {
        console.error('[rabbitmq] MESSAGING=amqp exige RABBITMQ_URL')
        return
    }

    const connection = await amqp.connect(env.RABBITMQ_URL)
    const channel = await connection.createChannel()
    await channel.assertExchange('neurofinance.dlx', 'fanout', { durable: true })
    await channel.assertQueue('neurofinance.dlq', { durable: true })
    await channel.bindQueue('neurofinance.dlq', 'neurofinance.dlx', '')
    await channel.assertExchange(env.RABBITMQ_EXCHANGE, 'topic', { durable: true })
    await channel.assertQueue(env.RABBITMQ_QUEUE_BACKEND, {
        durable: true,
        arguments: QUEUE_ARGS,
    })
    await channel.bindQueue(env.RABBITMQ_QUEUE_BACKEND, env.RABBITMQ_EXCHANGE, 'backend.#')
    await channel.prefetch(env.RABBITMQ_PREFETCH_BACKEND)

    await channel.consume(env.RABBITMQ_QUEUE_BACKEND, async (msg) => {
        if (!msg) {
            return
        }
        try {
            const result = await handleMessage(msg)
            channel.sendToQueue(
                msg.properties.replyTo,
                Buffer.from(JSON.stringify(result)),
                { correlationId: msg.properties.correlationId, contentType: 'application/json' },
            )
            channel.ack(msg)
        } catch (error) {
            console.error('[rabbitmq] falha ao processar mensagem', error)
            channel.nack(msg, false, false)
        }
    })

    console.log(`[rabbitmq] consumidor escutando ${env.RABBITMQ_QUEUE_BACKEND}`)
}

File: neuro-learning/messaging/consumer.py
import json
import os
import threading
import traceback

import pika


EXCHANGE = os.getenv("RABBITMQ_EXCHANGE", "neurofinance.topic")
QUEUE = os.getenv("RABBITMQ_QUEUE_LEARNING", "neuro.learning.rpc")
PREFETCH = int(os.getenv("RABBITMQ_PREFETCH_LEARNING", "1"))


def _dispatch(app, body: dict):
    method = (body.get("method") or "GET").upper()
    path = body.get("path") or "/"
    payload = body.get("payload")
    headers = body.get("headers") or {}
    authorization = headers.get("authorization") or headers.get("Authorization")

    request_headers = {}
    if authorization:
        request_headers["Authorization"] = authorization

    with app.test_client() as client:
        response = client.open(
            path,
            method=method,
            json=payload if method not in ("GET", "DELETE") else None,
            headers=request_headers,
        )

    parsed = response.get_json(silent=True)
    if parsed is None:
        parsed = response.get_data(as_text=True)
    return {"status": response.status_code, "body": parsed}


def _consume():
    from app import app

    url = os.getenv("RABBITMQ_URL")
    if not url:
        print("[rabbitmq] MESSAGING=amqp exige RABBITMQ_URL")
        return

    params = pika.URLParameters(url)
    params.heartbeat = 30
    connection = pika.BlockingConnection(params)
    channel = connection.channel()
    channel.exchange_declare(exchange="neurofinance.dlx", exchange_type="fanout", durable=True)
    channel.queue_declare(queue="neurofinance.dlq", durable=True)
    channel.queue_bind(queue="neurofinance.dlq", exchange="neurofinance.dlx", routing_key="")
    channel.exchange_declare(exchange=EXCHANGE, exchange_type="topic", durable=True)
    channel.queue_declare(
        queue=QUEUE,
        durable=True,
        arguments={"x-dead-letter-exchange": "neurofinance.dlx"},
    )
    channel.queue_bind(queue=QUEUE, exchange=EXCHANGE, routing_key="learning.#")
    channel.basic_qos(prefetch_count=PREFETCH)

    def on_message(ch, method, properties, body):
        try:
            request = json.loads(body.decode("utf-8"))
            result = _dispatch(app, request)
            if properties.reply_to:
                ch.basic_publish(
                    exchange="",
                    routing_key=properties.reply_to,
                    properties=pika.BasicProperties(
                        correlation_id=properties.correlation_id,
                        content_type="application/json",
                    ),
                    body=json.dumps(result).encode("utf-8"),
                )
            ch.basic_ack(delivery_tag=method.delivery_tag)
        except Exception:
            traceback.print_exc()
            ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)

    channel.basic_consume(queue=QUEUE, on_message_callback=on_message)
    print(f"[rabbitmq] consumidor escutando {QUEUE}")
    channel.start_consuming()


def start_learning_consumer():
    if os.getenv("MESSAGING", "http").lower() != "amqp":
        print("[rabbitmq] MESSAGING=http — consumidor do learning desligado")
        return
    thread = threading.Thread(target=_consume, name="rabbitmq-consumer", daemon=True)
    thread.start()
